File: js/src/toolsRead.ts
/**
 * Read-surface tools — docx_open / docx_outline / docx_read / docx_search —
 * with result shapes exactly per spec/tools/*.json and the §2a pins.
 * Errors are thrown as ToolError (spec/errors.json codes).
 */
import { ToolError } from "./errors.js";
import {
  type OutlineHeading,
  type OutlineTable,
  type ReadResult,
  type SearchMatch,
  outlineOf,
  paragraphHeadingLevel,
  projectionContext,
  readProjection,
  searchProjection,
} from "./projector.js";
import type { Session } from "./session.js";
import { nextTag } from "./xmlscan.js";

type ResponseFormat = "concise" | "detailed";

// ---------------------------------------------------------------------------
// docx_open
// ---------------------------------------------------------------------------

export interface DocxOpenArgs {
  path?: string | undefined;
  /** Base64-encoded .docx content (alternative to path). */
  bytes?: string | undefined;
  response_format?: ResponseFormat | undefined;
}

export interface DocxOpenResult {
  doc_id: string;
  summary: string;
  n_paragraphs: number;
  has_tracked_changes: boolean;
  has_comments: boolean;
}

function countNamed(xml: string, names: readonly string[]): Map<string, number> {
  const counts = new Map<string, number>(names.map((n) => [n, 0]));
  let i = 0;
  for (;;) {
    const t = nextTag(xml, i);
    if (!t) return counts;
    if (t.kind !== "end") {
      const c = counts.get(t.name);
      if (c !== undefined) counts.set(t.name, c + 1);
    }
    i = t.end;
  }
}

function plural(n: number, noun: string): string {
  return `${n} ${noun}${n === 1 ? "" : "s"}`;
}

/** Decode standard base64 via WHATWG `atob` (browsers and Node alike). */
function base64Bytes(b64: string): Uint8Array {
  let binary: string;
  try {
    binary = atob(b64);
  } catch (e) {
    throw new ToolError(
      "open_failed",
      `Cannot open: bytes is not valid base64 (${(e as Error).message}).`,
      ["Pass standard base64 (RFC 4648) of the .docx bytes."],
    );
  }
  const out = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) out[i] = binary.charCodeAt(i);
  return out;
}

export function docxOpen(session: Session, args: DocxOpenArgs): DocxOpenResult {
  if (args.path === undefined && args.bytes === undefined) {
    throw new ToolError("open_failed", "Cannot open: provide path or bytes.", [
      "Pass a filesystem path or base64-encoded bytes.",
    ]);
  }
  const doc =
    args.path !== undefined
      ? session.open(args.path)
      : session.open(base64Bytes(args.bytes as string));
  const xml = doc.documentXml();
  const ctx = projectionContext(doc);
  const index = doc.anchorIndex();
  let nParagraphs = 0;
  let nTables = 0;
  for (const e of index) {
    if (e.kind === "p") nParagraphs++;
    else nTables++;
  }
  const counts = countNamed(xml, ["w:sectPr", "w:ins", "w:del", "w:commentReference"]);
  // §2a title: first non-empty heading paragraph, else first non-empty
  // paragraph, else "Untitled".
  let title: string | null = null;
  let firstNonEmpty: string | null = null;
  for (const e of index) {
    if (e.kind !== "p" || !e.normalized) continue;
    if (paragraphHeadingLevel(xml, e.block, ctx.styles) !== null) {
      title = e.normalized;
      break;
    }
    if (firstNonEmpty === null) firstNonEmpty = e.normalized;
  }
  const summary =
    `${title ?? firstNonEmpty ?? "Untitled"} — ${plural(nParagraphs, "paragraph")}, ` +
    `${plural(counts.get("w:sectPr") ?? 0, "section")}, ${plural(nTables, "table")}`;
  return {
    doc_id: doc.id,
    summary,
    n_paragraphs: nParagraphs,
    has_tracked_changes: (counts.get("w:ins") ?? 0) + (counts.get("w:del") ?? 0) > 0,
    has_comments: (counts.get("w:commentReference") ?? 0) > 0,
  };
}

// ---------------------------------------------------------------------------
// docx_outline
// ---------------------------------------------------------------------------

export interface DocxOutlineArgs {
  doc_id: string;
  response_format?: ResponseFormat | undefined;
}

export interface DocxOutlineResult {
  outline: OutlineHeading[];
  tables: OutlineTable[];
}

export function docxOutline(session: Session, args: DocxOutlineArgs): DocxOutlineResult {
  return outlineOf(session.get(args.doc_id));
}

// ---------------------------------------------------------------------------
// docx_read
// ---------------------------------------------------------------------------

export interface DocxReadArgs {
  doc_id: string;
  anchor?: string | undefined;
  range?: string | undefined;
  window?: number | undefined;
  scope?: string | undefined;
  response_format?: ResponseFormat | undefined;
}

export function docxRead(session: Session, args: DocxReadArgs): ReadResult {
  const doc = session.get(args.doc_id);
  return readProjection(doc, {
    anchor: args.anchor,
    range: args.range,
    window: args.window,
    scope: args.scope,
  });
}

// ---------------------------------------------------------------------------
// docx_search
// ---------------------------------------------------------------------------

export interface DocxSearchArgs {
  doc_id: string;
  query: string;
  regex?: boolean | undefined;
  scope?: string | undefined;
  response_format?: ResponseFormat | undefined;
}

export interface DocxSearchResult {
  matches: SearchMatch[];
  n_matches: number;
}

export function docxSearch(session: Session, args: DocxSearchArgs): DocxSearchResult {
  const doc = session.get(args.doc_id);
  return searchProjection(doc, { query: args.query, regex: args.regex, scope: args.scope });
}

File: src/docxengine/_tools_read.py
"""Read-surface tools: docx_open, docx_outline, docx_read, docx_search.

Each function takes the :class:`~docxengine._session.Session` plus keyword
arguments named exactly as in ``spec/tools/<tool>.json`` and returns the result
object in exactly that schema's shape. Failures raise
:class:`~docxengine._errors.ToolError` with a ``spec/errors.json`` code.

``response_format`` is accepted everywhere per the schemas; in this phase the
``detailed`` view is identical to ``concise`` (run-level formatting lands later).
"""

from __future__ import annotations

import base64
import binascii
import re

from . import _projector
from ._errors import ToolError
from ._opc import Package
from ._session import OpenDocument, Session

_TRACKED_RE = re.compile(rb"<w:(?:ins|del)[ />]")
_COMMENT_REF_RE = re.compile(rb"<w:commentReference[ />]")
_SECT_PR_RE = re.compile(rb"<w:sectPr[ />]")

_Bytes = bytes  # the builtin, reachable where a schema-pinned parameter shadows it


def _has_tracked_changes(package: Package) -> bool:
    return _TRACKED_RE.search(package.part(package.main_document_part())) is not None


def _has_comments(package: Package) -> bool:
    return _COMMENT_REF_RE.search(package.part(package.main_document_part())) is not None


def _plural(n: int, noun: str) -> str:
    return f"{n} {noun}" if n == 1 else f"{n} {noun}s"


def _summary(package: Package, blocks: list[_projector.ProjectedBlock]) -> tuple[str, int]:
    """§2a: ``"{title} — {p} paragraphs, {s} sections, {t} tables"``."""
    paragraphs = [b for b in blocks if b.kind == "paragraph"]
    n_tables = sum(1 for b in blocks if b.kind == "table")
    n_sections = len(_SECT_PR_RE.findall(package.part(package.main_document_part())))
    title = next(
        (b.normalized for b in paragraphs if b.heading_level is not None and b.normalized),
        next((b.normalized for b in paragraphs if b.normalized), "Untitled"),
    )
    facts = [
        _plural(len(paragraphs), "paragraph"),
        _plural(n_sections, "section"),
        _plural(n_tables, "table"),
    ]
    return f"{title} — {', '.join(facts)}", len(paragraphs)


def docx_open(
    session: Session,
    *,
    path: str | None = None,
    bytes: str | None = None,  # noqa: A002 - wire name pinned by the tool schema
    response_format: str = "concise",
) -> dict[str, object]:
    """Open a .docx from a path or base64 bytes; returns the doc_id handle + summary."""
    if path is None and bytes is None:
        raise ToolError(
            "open_failed",
            "Cannot open document: provide either path or bytes.",
            ["Pass a filesystem path or base64-encoded .docx content."],
        )
    source: str | _Bytes
    if path is not None:
        source = path
    else:
        assert bytes is not None
        try:
            source = base64.b64decode(bytes, validate=True)
        except (binascii.Error, ValueError) as exc:
            raise ToolError(
                "open_failed",
                f"Cannot open <bytes>: invalid base64 content ({exc}).",
                ["Pass the .docx file content base64-encoded."],
            ) from exc
    doc = session.open_doc(source)
    summary, n_paragraphs = _summary(doc.package, _projector.project_body(doc.package))
    return {
        "doc_id": doc.doc_id,
        "summary": summary,
        "n_paragraphs": n_paragraphs,
        "has_tracked_changes": _has_tracked_changes(doc.package),
        "has_comments": _has_comments(doc.package),
    }


def docx_outline(
    session: Session, *, doc_id: str, response_format: str = "concise"
) -> dict[str, object]:
    """The heading tree and table list with anchors, style cascade resolved."""
    doc: OpenDocument = session.get(doc_id)
    return dict(_projector.project_outline(doc.package))


def docx_read(
    session: Session,
    *,
    doc_id: str,
    anchor: str | None = None,
    range: str | None = None,  # noqa: A002 - wire name pinned by the tool schema
    window: int = 0,
    scope: str = "body",
    response_format: str = "concise",
) -> dict[str, object]:
    """The Markdown projection of an anchor window, a range, or a whole story."""
    doc = session.get(doc_id)
    return dict(
        _projector.project_read(doc.package, anchor=anchor, range=range, window=window, scope=scope)
    )


def docx_search(
    session: Session,
    *,
    doc_id: str,
    query: str,
    regex: bool = False,
    scope: str = "body",
    response_format: str = "concise",
) -> dict[str, object]:
    """Coalesced-text search: matching anchors with snippets and heading context."""
    doc = session.get(doc_id)
    return dict(_projector.project_search(doc.package, query, regex=regex, scope=scope))
